Add conversions for 1x1 conv_2d to matmul

IanWood1 opened the pull request:
Convert 1x1 conv_2d to `linalg.matmul` ops when the HW dimensions are dynamic and convert `linalg.conv_2d_nhwc_hwcf` when the N dimension is not 1. No change to `linalg.conv_2d_nchw_fchw` currently (see linked issue for discussion). Matmul is simpler and easier for the compiler to understand, allowing for better optimizations.

IanWood1 committed:
Add conversion for NHWC

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>

IanWood1 committed:
Support dynamic H/W

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>

IanWood1 committed:
Add dep & adjust dispatch count

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>

Max191 commented on compiler/src/iree/compiler/GlobalOptimization/Convert1X1FilterConv2DToMatmul.cpp:
```diff
@@ -55,19 +57,11 @@ class Convert1x1FilterConvToMatmul : public OpRewritePattern<Conv2DOpType> {
 
     // TODO(ataei): Support conversion to linalg.batch_matmul.
-    if (inputShape[0] != 1)
+    if (isNCHW && inputShape[0] != 1)
```
See my comment on the linked issue: https://github.com/iree-org/iree/issues/18710#issuecomment-2407661454

I think we should generate a linalg.generic with multiple n contraction dimensions for NCHW.

IanWood1 replied:
Thank you, I'll take a look

Max191 reviewed (changes requested):
Can you support the NCHW case as well? I added a suggestion for how to do it.

Additional thought: It may be a reasonable idea to create a linalg.generic op with expanded H and W dims (so one M/N contraction dim for each of the N, H, and W convolution dims). Having fewer reshapes is probably better at GlobalOptimization level, but I am not sure if it is worth it yet. I think the collapse_shape on the H and W dimensions can typically be propagated through the producers/consumers, so it might not make a difference. You may have a better idea of which form is better for dispatch region formation.

Also, this additional thought is not a requirement for this PR, just an idea to try afterwards.

IanWood1 committed:
Generalize 1x1 Conv

Use linalg::generalizeNamedOp to generalize 1x1 conv and then remove the
unit extent affine symbols from input's affine map. Additionally, clean
up the test cases (since expand/extract is introduced). Just check that
the affine maps are correct.

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>

IanWood1 committed:
Add print-local-scope flag

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>

Max191 reviewed (commented):
I think we may not even need this pass anymore. Does it still work if you simply add a case for 1x1 filters here?
https://github.com/iree-org/iree/blob/c6056d197230161ea1403e88b5b8784d34e071a2/compiler/src/iree/compiler/GlobalOptimization/GeneralizeLinalgNamedOps.cpp#L40-L48

I would expect generalizing the op would be enough, and hopefully the indexing maps will simplify to contraction maps.

Max191 commented on .github/workflows/pkgci_regression_test.yml:
```diff
@@ -220,7 +220,7 @@ jobs:
             --goldentime-rocm-clip-ms 18.5 \
             --goldentime-rocm-vae-ms 337.0 \
-            --goldendispatch-rocm-unet 1551 \
+            --goldendispatch-rocm-unet 1540 \
```
Also, looks like these went down again (and some of the latencies too), so you can update these numbers.

IanWood1 committed:
Remove Convert1x1FilterConv2d files

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>

IanWood1 committed:
Add conv generalization to GeneralizeLinalgNamedOps

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>

IanWood1 committed:
Lower unet dispatch count and remove dep

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>

IanWood1 commented on compiler/src/iree/compiler/Preprocessing/Passes.cpp:
```diff
@@ -102,7 +102,6 @@ buildTransposeConvolutionPassPipeline(OpPassManager &passManager,
   FunctionLikeNest(passManager)
       .addPass(GlobalOptimization::createDetachElementwiseFromNamedOpsPass)
       .addPass(mlir::createLinalgNamedOpConversionPass)
-      .addPass(GlobalOptimization::createConvert1X1FilterConv2DToMatmulPass)
```
This might have some negative side effects

IanWood1 committed:
Merge branch 'main' into conv_to_matmul

MaheshRavishankar reviewed (changes requested):
I think changes to generalize is fine, but maybe keep the conv-> matmul conversion pass around (doesnt have to be used in the global optimization pass pipeline).

IanWood1 committed:
Revert "Remove Convert1x1FilterConv2d files"

This reverts commit ade26596ae9c1dbb5ebd29b2767dde72986931bb.

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>

Max191 commented on compiler/src/iree/compiler/GlobalOptimization/Convert1X1FilterConv2DToMatmul.cpp:
```diff
@@ -4,11 +4,19 @@
 // See https://llvm.org/LICENSE.txt for license information.
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
+#include <type_traits>
```
nit: Can you remove any of these new includes that aren't necessary

Max191 commented on compiler/src/iree/compiler/GlobalOptimization/GeneralizeLinalgNamedOps.cpp:
```diff
@@ -30,6 +30,24 @@ struct GeneralizeLinalgNamedOpsPass
 };
 } // namespace
 
+static bool is1x1FilterConv2dOp(linalg::LinalgOp convOp) {
```
You should also check that the strides of the conv are 1 (dilations are fine, since they do not affect 1x1 filter).

Also, maybe rename the function to something like `isConvFoldableToContraction`, and add some comments on the function describing the conditions for folding to a contraction.

Max191 reviewed (changes requested):
Mostly looks good, just one comment and one nit.

Also, I think you can remove the Convert1x1FilterConv2DToMatmul pass from GlobalOptimization/Passes.cpp now (or can be a follow up).

Do we need to look into the dispatch count regression for Unet? It is only 2 dispatches, and the perf seems to have improved, so maybe it is okay. Just make sure to update the golden dispatch counts once it is ready to land.

IanWood1 committed:
Address comments

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>

Max191 reviewed (approved):
Looks good! Just make sure to resolve the CI failures before landing (update golden dispatch counts).

IanWood1 committed:
Update sdxl CI

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>
